lib/inspec/attestations.rb

```diff
@@ -0,0 +1,152 @@
+        end
+      end
+    end
+
+    def self.validate_attestation_expiry(expiry, control_id)
+      # logic to check for expiry
+
+      if [Date, Time].include?(expiry.class) || (expiry.is_a?(String) && Time.parse(expiry).year != 0)
+        expiry = expiry.to_time if expiry.is_a? Date
+        expiry = Time.parse(expiry) if expiry.is_a? String
+        if expiry < Time.now # If the attestation expired, return - no attestation done
+          expiry_message = "Attestation expired on #{expiry}"
+          expiry_message
+        end
+      else
+        ui = Inspec::UI.new
+        ui.error("Unable to parse attestation expiration date '#{expiry}' for control #{control_id}")
+        ui.exit(:usage_error)
+      end
+    rescue => e
+      ui = Inspec::UI.new
+      ui.error("Unable to parse attestation expiration date '#{expiry}' for control #{control_id}. Error: #{e.message}")
+      ui.exit(:usage_error)
+    end
+
+    def self.attestation_check(expiry_message, attestation_data, control_id)
+      # logic to update enhanced outcome status
+      status, msg = nil
+      if %w{passed failed}.include? attestation_data["status"]
+        if expiry_message
+          status = "failed"
+          msg = "Control not attested : #{expiry_message}"
+        else
+          # use justification and evidence url to show information in msg
+          attestation_message = attestation_data["justification"] || attestation_data["explanation"] || ""
+
+          unless attestation_data["evidence_url"].blank?
+            if attestation_message.blank?
+              attestation_message = "Evidence URL: #{attestation_data["evidence_url"]}"
+            else
+              attestation_message += " | Evidence URL: #{attestation_data["evidence_url"]}"
+            end
+          end
+
+          status = attestation_data["status"]
+          if attestation_message.blank?
+            msg = "Control Attested : No justification provided."
+          else
+            msg = "Control Attested : #{attestation_message}"
+          end
+        end
+      else
+        if attestation_data["status"].blank?
+          Inspec::Log.warn "No attestation status for control #{control_id}. Use 'passed' or 'failed'."
+        else
+          Inspec::Log.warn "Invalid attestation status '#{attestation_data["status"]}' for control #{control_id}. Use 'passed' or 'failed'."
+        end
+        return nil
+      end
+      [status, msg]
+    end
+
+    def self.determine_expiry(attestation_data, control_id)
+      if attestation_data["expiration_date"]
+        attestation_data["expiration_date"]
+      elsif !attestation_data["updated"].blank? && !attestation_data["frequency"].blank?
+        # logic to find expiration date using frequency and updated date.
+        updated_last = attestation_data["updated"]
+        begin
+          if updated_last.is_a?(Date) || (updated_last.is_a?(String) && Date.parse(updated_last).year != 0)
```

Comment:
Maybe you can write this whole bunch of if and else in another method to make it cleaner.

Reply:
Sure, let me think about how I can logically separate those lines since I think they are part of the logic used for fetching the expiration date.